pkg/kubelet/kubelet_pods.go

```diff
@@ -1819,11 +1819,15 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 		statuses[container.Name] = status
 	}
 
+	// Copy the slice before sorting it
+	containerStatusesCopy := make([]*kubecontainer.Status, len(podStatus.ContainerStatuses))
+	copy(containerStatusesCopy, podStatus.ContainerStatuses)
```

Comment:
I'm wondering if rw locking the slice would be better than making a whole copy. :thinking:

Reply:
Thanks for the review ! A couple of questions regarding a lock-based approach:
- The function / code path is supposed to be purely transformative and not alter the kubelet State (I suppose, including the podCache), should we change that assumption ? https://github.com/kubernetes/kubernetes/blob/v1.20.4/pkg/kubelet/kubelet_pods.go#L1540-L1542 
It would also potentially alter the assumption that callers should not modify objects from the pod cache https://github.com/kubernetes/kubernetes/blob/v1.20.4/pkg/kubelet/container/cache.go#L81-L88
- Where should the lock live ? Should we add an attribute `ContainerStatusesLock` to the pod status https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container/runtime.go#L290-L304 / or higher on the podCache level ?

Reply:
> * The function / code path is supposed to be purely transformative and not alter the kubelet State (I suppose, including the podCache), should we change that assumption ? https://github.com/kubernetes/kubernetes/blob/v1.20.4/pkg/kubelet/kubelet_pods.go#L1540-L1542

I think so, since the assumption seems wrong with the `Sort()` call. We could also assume that the status is already sorted and move the sorting code to another code path, but this way we would still require the lock from what I can see.

>   It would also potentially alter the assumption that callers should not modify objects from the pod cache https://github.com/kubernetes/kubernetes/blob/v1.20.4/pkg/kubelet/container/cache.go#L81-L88
> * Where should the lock live ? Should we add an attribute `ContainerStatusesLock` to the pod status https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container/runtime.go#L290-L304 / or higher on the podCache level ?

A higher lock would mean that we lock the whole `PodStatus`, right? I personally would prefer locking only the structure we need, which means the lock would be part of `PodStatus`.